right/src/config_parser/parse_module_config.c

```diff
@@ -0,0 +1,125 @@
+#include "parse_module_config.h"
+#include "basic_types.h"
+#include "config_globals.h"
+#include <stdint.h>
+#include "config.h"
+#include "parse_config.h"
+#include "layer.h"
+#include "module.h"
+#include "mouse_controller.h"
+
+static parser_error_t parseNavigationModes(config_buffer_t *buffer, module_configuration_t* moduleConfiguration)
+{
+    // parse navigation modes
+    for (uint8_t layerId = 0; layerId < LayerId_RegularCount; layerId++) {
```

Comment:
@kareltucek I am writing the module config parser of the Agent. I am confused because  the `Base` layer id of the Agent is `255`, but the firmware uses `0`
Agent Layer enum definition: https://github.com/UltimateHackingKeyboard/agent/blob/master/packages/uhk-common/src/config-serializer/config-items/layer-name.ts
Firmware Layer enum definition: https://github.com/UltimateHackingKeyboard/firmware/blob/master/right/src/layer.h#L14

First I did not understand how it had not cause any problem. But I found the [parseLayerId](https://github.com/UltimateHackingKeyboard/firmware/blob/master/right/src/config_parser/parse_keymap.c#L180) function that has made the transformation. 

The `parseNavigationModes` uses the internal `LayerId` of the firmware. If I implement the above logic then I have to create the same transformation logic which is in the `parseLayerId` function. 

In order to be consistent the firmware should use the transformation logic in the `parseNavigationModes` function too.

What do you think?

Reply:
I see. I didn't realize this inconsistency.

Well, this will need special handling somewhere anyways, as we clearly don't want to serialize a 256 element array just in order to have base layer at position 255. 

Options I see:
- serialize it as a sequence with base layer first (i.e., to satisfy this code)
- serialize it as a sequence with base layer last (to conform to the order of the agent enum)
- serialize it as a dictionary of key value pairs.

The conceptually cleanest solution seems to be the third one. But from practical standpoint I don't think this is an important decision, and any of the three is equally good. I slightly lean towards leaving it as is - i.e., the first option - but will adapt to any one you should choose.

Reply:
Before I answer, I would like to ask. Would the `navigationMode` be a structure in the future instead of primitive value?

currently:
```
navigationMode[Layer_Base] = Cursor
```

complex version
```
navigationMode[Layer_Base] = {
  navigationMode: Cursor
  propertyA: Something
}
```

If it could be a complex type then we will serialise it as an array and for flexibility it is worth serialising the `LayerId` as part of the structure. Should we prepare for the complex type version?

```
navigationModes = [
  {
     id: layerId,
     navigationMode: Cursor
     propertyA: Something
  }
]
```


Just now realised does not matter what the structure of the `navigationMode` is/will be, because serialisation, transformation logic should be moved to the Agent side. So the `serialize it as a sequence with base layer first (i.e., to satisfy this code)` version is logical and  the [parseLayerId](https://github.com/UltimateHackingKeyboard/firmware/blob/master/right/src/config_parser/parse_keymap.c?rgh-link-date=2024-02-11T20%3A44%3A49Z#L180) function needs to be refactored.

Reply:
I prefer the first option as that'd lead to the cleanest solution from a serialization standpoint.

But I'm curious what may be the reason for the current serialization scheme because maybe I'm missing something.

Reply:
> Would the navigationMode be a structure in the future instead of primitive value?

I don't see any future need for it to be a structure at this moment.

> Just now realised does not matter what the structure of the navigationMode is/will be, because serialisation, transformation logic should be moved to the Agent side. So the serialize it as a sequence with base layer first (i.e., to satisfy this code) version is logical and the [parseLayerId](https://github.com/UltimateHackingKeyboard/firmware/blob/master/right/src/config_parser/parse_keymap.c?rgh-link-date=2024-02-11T20%3A44%3A49Z#L180) function needs to be refactored.

I am confused. Do I understand correctly that you propose to make the `LayerId_Base = 0` universal in the serialization format, and therefore basically remove parseLayerId from firmware code? (Or more precisely convert to a backwards-compatible format.)

> But I'm curious what may be the reason for the current serialization scheme because maybe I'm missing something.

I don't understand? 

If you are asking as to the idea behind serialized base layer having id of 255, I am baffled by that too... 

I guess the original reason was that layer ids were used to toggle/hold layers, and thus base layer id was not needed in the enum at all.

Reply:
I reviewed the PR that introduced the extra layers handling.  https://github.com/UltimateHackingKeyboard/agent/pull/1548/files#diff-2e6e10c7a57591441db51b02684782c9706fe152f7accc8ddac69a78fac02f64R4

The `LayerName` enum already existed without the `base` layer value.
```typescript
export enum LayerName {
     mod,
     fn,
     mouse
 }
```
We added the `base` layer as the last item for backward compatibility.

Reply:
> I am confused. Do I understand correctly that you propose to make the `LayerId_Base = 0` universal in the serialization format, and therefore basically remove parseLayerId from firmware code?

Yes.

> (Or more precisely convert to a backwards-compatible format.)

Do you mean backwards-incompatible?

> I guess the original reason was that layer ids were used to toggle/hold layers, and thus base layer id was not needed in the enum at all.

I agree that this was probably the reason.

Reply:
> I am confused. Do I understand correctly that you propose to make the LayerId_Base = 0 universal in the serialization format, and therefore basically remove parseLayerId from firmware code? (Or more precisely convert to a backwards-compatible format.)

Yes

Reply:
> Do you mean backwards-incompatible? 

Well, the new version does not need explicit conversion at all as the proposed serialized layer id is same as internal firmware layer id. So the only point of leaving there the function at all is for backward-compatibility purposes. So I mean to actually make it handle backward compatibility.

Reply:
Ok! I will modify the code correspondingly.

Reply:
Done.